bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -183,80 +318,90 @@ impl AuthenticationService {
+        };
+
+        let Some(oidc_configuration) = &self.oidc_configuration else {
+            return Err(AuthenticationError::OidcMetadataMissing);
+        };
+
+        let redirect_url = Url::parse(&oidc_configuration.redirect_uri)
+            .map_err(|_e| AuthenticationError::OidcMetadataInvalid)?;
+
+        let oidc = client.inner.oidc();
+
+        RUNTIME.block_on(async {
+            self.configure_oidc(&oidc, authentication_server, oidc_configuration).await?;
+
+            let data = oidc.login(redirect_url, None)?.build().await?;
+
+            Ok(Arc::new(OidcAuthenticationData { url: data.url, state: data.state }))
+        })
     }
 
-    /// Restore an existing session on the current homeserver using an access
-    /// token issued by an authentication server.
-    /// # Arguments
-    ///
-    /// * `token` - The access token issued by the authentication server.
-    ///
-    /// * `device_id` - The device ID that the access token was scoped for.
-    pub fn restore_with_access_token(
+    pub fn login_with_oidc_callback(
         &self,
-        token: String,
-        device_id: String,
+        authentication_data: Arc<OidcAuthenticationData>,
+        callback_url: String,
     ) -> Result<Arc<Client>, AuthenticationError> {
         let Some(client) = self.client.read().unwrap().clone() else {
             return Err(AuthenticationError::ClientMissing);
         };
 
-        // Restore the client and ask the server for the full user ID as this
-        // could be different from the username that was entered.
-        let discovery_user_id = UserId::parse("@unknown:unknown")
-            .map_err(|e| AuthenticationError::Generic { message: e.to_string() })?;
-        let device_id: OwnedDeviceId = device_id.as_str().into();
+        let oidc = client.inner.oidc();
 
-        let discovery_session = Session {
-            meta: SessionMeta { user_id: discovery_user_id, device_id: device_id.clone() },
-            tokens: SessionTokens { access_token: token.clone(), refresh_token: None },
-        };
+        let url =
+            Url::parse(&callback_url).map_err(|_| AuthenticationError::OidcCallbackUrlInvalid)?;
 
-        client.restore_session_inner(discovery_session)?;
-        let whoami = client.whoami()?;
+        let response = AuthorizationResponse::parse_uri(&url)
+            .map_err(|_| AuthenticationError::OidcCallbackUrlInvalid)?;
 
-        // Create the actual client with a store path from the user ID.
-        let homeserver_url = client.homeserver();
-        let session = Session {
-            meta: SessionMeta { user_id: whoami.user_id.clone(), device_id },
-            tokens: SessionTokens { access_token: token, refresh_token: None },
+        let code = match response {
+            AuthorizationResponse::Success(code) => code,
+            AuthorizationResponse::Error(err) => {
+                // TODO: This doesn't work, cancel doesn't include a description it seems.
```

Comment:
`error` is None in this instance, looking into why it is, but will see if it can be re-written using a typed error instead of matching strings.

Reply:
Fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/2412/commits/be9dbe1ef612050e806c17a1172494064676ef0a